lib/main.js

```diff
@@ -189,99 +202,186 @@ function displayMessage(m_title, message, type) {
+
+  if (action === "clear_queue")
+    actions_queue.push({ 'action': "clear", 'handler': "handleComplete", 'url': rurl, "file": "" });
+
+  if (action === "synchronize")
+    actions_queue.push({ 'action': "synchronize", 'handler': "handle_get_items", 'url': rurl, "file": "" });
+
+  displayMessage('Sending', 'Sending to Kodi...', 'info');
+
+  if (is_queue_empty)
+    request_xbmc(actions_queue.shift());
+}
+
+function request_xbmc(request) {
+  var play_item = {
     "jsonrpc":"2.0",
     "method":"Player.Open",
     "params": {
       "item":{
-        "file":fileurl
+        "position": plists.get_size(),
+        "playlistid": plists.playlist_id
       }
-    },
-    "id": 1
+    }
   };
-  displayMessage('Sending', 'Sending to Kodi...', 'info');
+
+  var add_playlist_item = {
+    "jsonrpc": "2.0",
+    "method": "Playlist.Add",
+    "params": {
+      "playlistid": plists.playlist_id,
+      "item": {
+        "file": request.file
+      }
+    }
+  };
+
+  var clear_playlist = {
+    "jsonrpc": "2.0",
+    "method": "Playlist.Clear",
+    "params": {
+      "playlistid": plists.playlist_id
+    }
+  };
+
+  var get_playlist_items = {
+    "jsonrpc": "2.0",
+    "method": "Playlist.GetItems",
+    "params": {
+      "playlistid": plists.playlist_id
+    }
+  };
+
+  var action = {};
+
+  // Determine what to do depending on the action.
+  if (request.action === "play")
+    action = play_item;
+
+  if (request.action === "add_queue")
+    action = add_playlist_item;
+
+  if (request.action === "clear_queue")
+    action = clear_playlist;
+
+  if (request.action === "synchronize")
+    action = get_playlist_items;
+
+  action.id = 1;
   Request({
-    url: rurl,
-    content: JSON.stringify(senddata),
-    //TODO: Playlists
-    //content: '{"jsonrpc": "2.0", "method": "Playlist.Add", "params":{ "playlistid":1, "item": {"file" : "' + fileurl + '" } }, "id" : 1}',
+    url: request.url,
+    content: JSON.stringify(action),
     contentType: 'application/json',
     onError: function (resp) {
-      //console.log(resp);
+      console.log(resp);
     },
-    onComplete: handleComplete
+    onComplete: function (handler) {
+      var h = handler;
+
+      return function (resp) {
+        handle_response(resp, h);
+
+        if (actions_queue.length > 0)
+          request_xbmc(actions_queue.shift());
+      };
+    }(request.handler)
   }).post();
 }
-//Handle return from XBMC
-function handleComplete(resp) {
-  //console.log(resp.status);
+
+function handle_get_items(resp) {
+  resp = JSON.parse(resp);
+  plists.resize(resp.result.limits.total);
+};
+
+function handle_response(resp, handler) {
   if (resp.status == 200) {
-    if (resp.json && resp.json.result) {
-      if (resp.json.result == 'OK') {
-        displayMessage('Success', 'Sent to Kodi', 'ok');
-        return;
-      }
+    if (resp.json) {
+      eval(handler + '(JSON.stringify(resp.json))');
```

Comment:
It seems that the json can't have any function pointer in its structure, this is the only way I found to transport handlers. Maybe an other way would be to declare a function pointer array and store indexes or enums in the json, it would be safer I think.